internal/federationdomain/resolvedprovider/resolved_provider.go

```diff
@@ -1,30 +1,176 @@
+	GetDisplayName() string
+
+	// GetProvider returns a representation of the upstream identity provider custom resource related to this
+	// identity provider for the FederationDomain, e.g. the OIDCIdentityProvider.
+	GetProvider() upstreamprovider.UpstreamIdentityProviderI
+
+	// GetSessionProviderType returns the type of session created by this identity provider.
+	GetSessionProviderType() psession.ProviderType
+
+	// GetIDPDiscoveryType returns the type of this identity provider, as shown by the IDP discovery endpoint.
+	GetIDPDiscoveryType() v1alpha1.IDPType
+
+	// GetIDPDiscoveryFlows returns the supported flows of this identity provider,
+	// as shown by the IDP discovery endpoint.
+	GetIDPDiscoveryFlows() []v1alpha1.IDPFlow
+
+	// GetTransforms returns the compiled version of the identity transformations and policies configured on the
+	// FederationDomain for this identity provider.
+	GetTransforms() *idtransform.TransformationPipeline
+
+	// CloneIDPSpecificSessionDataFromSession should reach into the provided session and return a clone
+	// of the field which is specific to the upstream identity provider type. If the session's field is
+	// nil, then return nil.
+	// Refer to the fields of psession.CustomSessionData whose types are specific to an identity provider type.
+	CloneIDPSpecificSessionDataFromSession(session *psession.CustomSessionData) interface{}
+
+	// ApplyIDPSpecificSessionDataToSession assigns the IDP-specific portion of the session data into a session.
+	// The IDP-specific session data provided to this function will be from an Identity that was returned by
+	// one of the other functions of this interface, so an implementation of this function can make safe
+	// assumptions about the type of idpSpecificSessionData for casting, based upon how it chooses to return
+	// IDPSpecificSessionData in Identity structs. If the given session already has any IDP-specific session
+	// data, it should be overwritten by this function.
+	ApplyIDPSpecificSessionDataToSession(session *psession.CustomSessionData, idpSpecificSessionData interface{})
+
+	// UpstreamAuthorizeRedirectURL returns the URL to which the user's browser can be redirected to continue
+	// the downstream browser-based authorization flow. Returned errors should be of type fosite.RFC6749Error.
+	UpstreamAuthorizeRedirectURL(state *UpstreamAuthorizeRequestState, downstreamIssuerURL string) (string, error)
+
+	// Login performs auth using a username and password that was submitted by the client, without a web browser.
+	// This function should authenticate the user with the upstream identity provider, extract their upstream
+	// identity, and transform it into their downstream identity.
+	// The groupsWillBeIgnored parameter will be true when the returned groups are going to be ignored by the caller,
+	// in which case this function may be able to save some effort by avoiding getting the user's upstream groups.
+	// Returned errors should be of type fosite.RFC6749Error.
```

Comment:
2 thoughts:

- Should "Login" be between UpstreamAuthorizeRedirectURL and HandleCallback when these two functions are actually related?
- Given "Login" and "HandleCallback" both do login, should we consider a name for Login that is more specific?

Reply:
Sounds good. Updated in latest commit.